File: DiagonalTraverse-498.py
# Time Complexity : O(m*n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach 

#Have a direction flag to change the workflow and check for boundary conditions. Rest is just changing row and column index

class Solution(object):
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if matrix == None or len(matrix) == 0 or len(matrix[0]) == 0:
            return []
        
        m,n = len(matrix), len(matrix[0])
        dirs = 1
        ans = [0 for _ in range(m*n)]
        index, i, j = 0, 0, 0
        
        while index < m*n:
            ans[index] = matrix[i][j]
            index += 1
            
            if dirs == 1:
                if j == n - 1:
                    dirs = -1
                    i += 1
                elif i == 0:
                    dirs = -1
                    j += 1
                else:
                    i -= 1
                    j += 1
            else:
                if i == m - 1:
                    dirs = 1
                    j += 1
                elif j == 0:
                    dirs = 1
                    i += 1
                else:
                    j -= 1
                    i += 1
        return ans

File: diagonalTraverse_498.cpp
// Time Complexity : O(MN)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        if(mat.empty() || mat.size() == 0){
            vector<int> x;
            return x;
        }
        int r=0, c=0, i=0,dir=1;
        int m = mat.size();
        int n = mat[0].size();
        vector<int> arr(m*n);
        while(i < m*n){
            arr[i] = mat[r][c];
            if(dir == 1 ){
                if(c == n-1){
                    r++;
                    dir = -1;
                }
                else if(r == 0){
                    c++;
                    dir = -1;
                }
                else{
                    r--;
                    c++;
                }
            }
            else{
                if(r == m-1){
                    c++;
                    dir = 1;
                }
                else if(c == 0){
                    r++;
                    dir = 1;
                }
                else{
                    c--;
                    r++;
                }
            }
            i++;
        }
        return arr;
    }
};
